opal/runtime/opal_init.c

```diff
@@ -193,11 +195,10 @@ int opal_init(int *pargc, char ***pargv)
         return opal_init_error("opal_init framework open", ret);
     }
 
-    /* Intitialize Accelerator framework
+    /* select Accelerator framework component
      * The datatype convertor code has a dependency on the accelerator framework
      * being initialized. */
-    ret = mca_base_framework_open(&opal_accelerator_base_framework, 0);
```

Comment:
I don't understand how this change fixes anything.  `mca_base_framework_open_list()` is literally just a for loop around the framework list calling `mca_base_framework_open()` on each item in the list.  It has the same flag argument, so there's no change there.  I don't think this was the important change.

I thought it might be a different behavior in the registered cleanup argument, but reading the code for `mca_base_framework_close_list`, it is just calling `mca_base_framework_close()`.  I think the real problem is not any of the list usage stuff, but that the registered cleanup function was/is calling the finalize function directly rather than calling `mca_base_framework_close`, which was a pretty silly mistake on our part in the review.  It kind of makes me wonder where @wckzhang found that cleanup example and if we don't have other issues lurking elsewhere.

Either way, this patch is wrong.  Either the `opal_finalize_register_cleanup` on line 203 needs to be removed or the whole patch could become changing that to `opal_finalize_register_cleanup_arg(mca_base_framework_close, opal_accelerator_base_framework)`.  The first obviously seems simpler, but one of the two needs to happen.

Reply:
before the patch here's the dlopening/closing activity related to accelerator framework:

```

calling MPI_INit
dlopnening /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_rocm
dlopnening /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_null

calling MPI_Finalize
dlclosing /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_rocm
dlclosing /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_null

calling MPI_Session_init
hit segfault in the accelerator framework selection function as there are pointers to memory which has been munmapped.
```
behavior with the patch
```
MPI_Init
dlopnening /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_rocm
dlopnening /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_null
MPI_Finalize
dlclosing /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_rocm
dlclosing /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_null
MPI_Session_init
dlopnening /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_rocm
dlopnening /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_null
MPI_Session_finalize
dlclosing /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_rocm
dlclosing /usr/projects/artab/users/hpp/ompi3/test_install/lib/openmpi/mca_accelerator_null
```
a framework has to  be closed in opal_finalize when using enable-mca-dso otherwise a reopen of the framework is a no op owing to a check at line 173 of mca_base_framework.c but the component *.so's were unloaded as part of the MPI_Finalize.  Without the patch the accelerator framework was not being closed.

Reply:
yep I agree about removing the accelerator finalize call. 
@wckzhang do you recall what motivated having the accelerator component finalize handled in a separate function rather than in the close of the component?  Is there some cuda behavior that motivated this?

Reply:
I just thought that was how we do cleanup

Reply:
I think before I added that, I wasn't seeing the cleanup actually happening, but it's been a while so I don't remember the explicit behavior.

Reply:
ah okay this makes sense as the code wasn't closing the framework.  I was just thinking cuda may be weird and you needed to use a different approach.

Reply:
@wckzhang i did more checking and actually the approach you were taking for cleanup in the accelerator framework is the better one, actually its necessary.

Reply:
@hppritcha I don't understand that.  `mca_base_framework_close()` is still the right answer; calling `<framework>.finalize()` directly leaves the base framework in a not-clean state.

Reply:
actually the accelerator framework was mostly doing the right thing.  As described in section IIIB of [10.1109/CLUSTER.2019.8891002](https://doi.org/10.1109/CLUSTER.2019.8891002) the complex interlinking of framework dependencies that existed prior to sessions was replaced with a cleanup callback framework. 

@hjelmn can provide more details if needed.

What is the not-clean state that the accelerator framework is being left in?

Reply:
The code calls:

```
opal_finalize_register_cleanup(opal_accelerator_base_selected_component.accelerator_finalize);
```

instead of the correct:

```
opal_finalize_register_cleanup_arg(mca_base_framework_close, opal_accelerator_base_framework);
```

So while the component gets cleaned up, the framework itself does not.

Reply:
Why would ```opal_finalize_register_cleanup_arg(mca_base_framework_close, opal_accelerator_base_framework)``` be needed? If it is in the list of frameworks (which it should be if you want it always opened) then it will get called automatically when the last call to finalize is made.

I plan to take a closer look at this PR and the existing code tomorrow. Will probably have more comments then.

Reply:
@hjelmn we've gone around so long that it doesn't matter.  But the current PR is wrong and needs to be fixed.  Today, on line 212, it calls `opal_finalize_register_cleanup(opal_accelerator_base_selected_component.accelerator_finalize);`.  This line just needs to be removed.

I was trying to make the point to Howard that adding the accelerator framework to the list of frameworks wasn't strictly necessary to fixing the issue, because I was pushing on how we missed this issue in the original CR.  That just overcomplicated things.  If the patch is extended to remove line 212, I'm happy to ship this patch.  Until then, the patch is wrong, because we're calling the component's `accelerator_finalize()` function twice.  Once from the base cleanup function that is invoked by adding the framework to the opal framework list and once explicitly on line 212.